subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java

```diff
@@ -53,37 +55,47 @@ public static VariantSelectionResult selectVariantsUsingAttributeMatching(Immuta
 
     private static VariantSelectionResult selectVariantsUsingAttributeMatching(ImmutableAttributes consumerAttributes, Collection<? extends Capability> explicitRequestedCapabilities, ComponentGraphResolveState targetComponentState, AttributesSchemaInternal consumerSchema, List<IvyArtifactName> requestedArtifacts, AttributeMatchingExplanationBuilder explanationBuilder) {
         ComponentGraphResolveMetadata targetComponent = targetComponentState.getMetadata();
+        AttributeMatcher attributeMatcher = consumerSchema.withProducer(targetComponent.getAttributesSchema());
+
         Optional<List<? extends VariantGraphResolveMetadata>> variantsForGraphTraversal = targetComponent.getVariantsForGraphTraversal();
-        List<? extends VariantGraphResolveMetadata> consumableVariants = variantsForGraphTraversal.or(ImmutableList.of());
-        AttributesSchemaInternal producerAttributeSchema = targetComponent.getAttributesSchema();
-        AttributeMatcher attributeMatcher = consumerSchema.withProducer(producerAttributeSchema);
-        ConfigurationGraphResolveMetadata fallbackConfiguration = targetComponent.getConfiguration(Dependency.DEFAULT_CONFIGURATION);
-        if (fallbackConfiguration != null && !fallbackConfiguration.isCanBeConsumed()) {
-            fallbackConfiguration = null;
-        }
-        ModuleVersionIdentifier versionId = targetComponent.getModuleVersionId();
-        if (!consumableVariants.isEmpty()) {
-            ImmutableList<VariantGraphResolveMetadata> variantsProvidingRequestedCapabilities = filterVariantsByRequestedCapabilities(targetComponent, explicitRequestedCapabilities, consumableVariants, versionId.getGroup(), versionId.getName(), true);
-            if (variantsProvidingRequestedCapabilities.isEmpty()) {
-                throw new NoMatchingCapabilitiesException(targetComponent, explicitRequestedCapabilities, consumableVariants);
+        boolean variantAware = variantsForGraphTraversal.isPresent();
+
+        // Fallback to the default configuration if there are no variants or if variant aware resolution is not supported.
+        if (!variantAware || variantsForGraphTraversal.get().isEmpty()) {
+            ConfigurationGraphResolveMetadata fallbackConfiguration = targetComponent.getConfiguration(Dependency.DEFAULT_CONFIGURATION);
+            if (fallbackConfiguration != null &&
+                fallbackConfiguration.isCanBeConsumed() &&
+                attributeMatcher.isMatching(fallbackConfiguration.getAttributes(), consumerAttributes)
+            ) {
+                return singleVariant(variantAware, ImmutableList.of(fallbackConfiguration));
             }
-            consumableVariants = variantsProvidingRequestedCapabilities;
+
+            AttributeDescriber describer = DescriberSelector.selectDescriber(consumerAttributes, consumerSchema);
+            throw new NoMatchingConfigurationSelectionException(describer, consumerAttributes, attributeMatcher, targetComponent, variantAware);
```

Comment:
💭 Would it be useful to expand the message in this exception to add something about even the default configuration not matching in this case?

Reply:
I've pulled in the terrible variant selection exception message issue into 8.2. This will likely be updated when working on that issue.